client/ConnectionThread.py

```diff
@@ -21,61 +36,113 @@ def run(self):
                         self.listen()
                 else:
                     break
-            self.conn.settimeout(3.0)
+            if self.conn is not None:
+               self.conn.settimeout(2.0)
             while not self.handler_status.to_kill:
-                self.send_counter()
-                time.sleep(1.0)
-                self.get_counter()
-                self.handler_status.msg += 1
+                if self.end_game:
+                    pass # send end_game message
+                if not self.last_package:
+                    self.send_counter()
+                    self.get_counter()
+                    if self.numbers_of_errors >= self.max_numbers_of_errors:
+                        self.handler_status.to_kill = 1
+                        self.handler_status.is_connected = 'connection_error'
+                        self.disconnected = True
+                    time.sleep(0.05)
         finally:
+            self.logger.debug("Conn thread closed")
             self.kill()
 
+    def set_message_to_send(self, message):
+        while self.lock.locked():
+            time.sleep(0.05)
+        self.handler_status.msg_to_send = message
+
+    def get_received_message(self):
+        while self.lock.locked():
+            time.sleep(0.05)
+        msg = self.handler_status.received_msg
+        return msg
+
+    def is_connected(self):
+        return self.established_connection
+
+
     def send_counter(self):
-        self.lock.acquire()
-        for i in range(5):
+        if not self.established_connection:
+            for i in range(5):
+                try:
+                    self.lock.acquire()
+                    if not self.established_connection:
+                        msg = f"{self.handler_status.host_username} -> {self.handler_status.player_name}:{self.handler_status.msg_to_send}"
+                    else:
+                        msg = self.handler_status.msg_to_send
+                    self.lock.release()
+                    sent = self.conn.send(str(msg).encode())
+                    break
+                except socket.error:
+                    if self.lock.locked():
+                        self.lock.release()
+                    if i == 5:
+                        self.logger.debug("Couldn't send message. Receiver must've closed connection!")
+                        self.handler_status.is_connected = "connection_error"
```

Comment:
i'd suggest self.lock.release() in finally block. More pythonic way

Reply:
To be honest, I'm not sure why, but this suggestion results in some more response time needed. That's why I've done it this way. After discussion, we decided to leave it as it is.